### PR TITLE
Misc Fixes vor Windows / CUDA 10.0

### PR DESCRIPTION
**Pull request by jvschroeder**
Fixed some isssues I encountered while trying to compile the examples on Windows using CUDA 10.0

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -119,7 +119,7 @@ add_executable(version_management EXCLUDE_FROM_ALL examples/by_runtime_api_modul
 add_executable(error_handling EXCLUDE_FROM_ALL examples/by_runtime_api_module/error_handling.cu)
 add_executable(device_management EXCLUDE_FROM_ALL examples/by_runtime_api_module/device_management.cpp)
 add_executable(execution_control EXCLUDE_FROM_ALL examples/by_runtime_api_module/execution_control.cu)
-target_compile_options(execution_control PRIVATE -rdc true)
+#target_compile_options(execution_control PRIVATE -rdc true)
```
Comment (eyalroz) [inline CMakeLists.txt]: IIRC, this is necessary on non-Windows machines, I can't just remove it.

File: cmake/Modules/HandleCUDAComputeCapability.cmake [hunk 2]
```diff
@@ -8,24 +8,44 @@
 # those without the CUDA_ prefix
 
 if (NOT CUDA_TARGET_COMPUTE_CAPABILITY)
-	if("$ENV{CUDA_SM}" STREQUAL "")
-		set(ENV{CUDA_INCLUDE_DIRS} "${CUDA_INCLUDE_DIRS}")
-		set(ENV{CUDA_CUDART_LIBRARY} "${CUDA_CUDART_LIBRARY}")
-		set(ENV{CMAKE_CXX_COMPILER} "${CMAKE_CXX_COMPILER}")
-		execute_process(COMMAND bash -c "${CMAKE_CURRENT_SOURCE_DIR}/scripts/get_cuda_sm.sh"  OUTPUT_VARIABLE CUDA_TARGET_COMPUTE_CAPABILITY_) 
+	if(WIN32)
+		find_package(CUDA)
```
Comment (eyalroz) [inline cmake/Modules/HandleCUDAComputeCapability.cmake]: CUDA is now an integral part of CMake, we don't use find_package on it. All CUDA-related variables are built-in.

File: cmake/Modules/HandleCUDAComputeCapability.cmake [hunk 3]
```diff
@@ -8,24 +8,44 @@
...
+		if (CUDA_FOUND)
+			set(CUDAFILE ${CMAKE_CURRENT_SOURCE_DIR}/examples/check_cuda.cu)
```
Comment (eyalroz) [inline cmake/Modules/HandleCUDAComputeCapability.cmake]: It seems you've developed a facility for detecting the CUDA SM on Windows. Please:

1. Make that a separate PR
2. Put it in a CMake module
3. If you can spare the time - See how much of the Unix-oriented shell script can be discarded in favor of what you do in CMake, so that the code paths will not look so different.

File: cmake/Modules/HandleCUDAComputeCapability.cmake [hunk 4]
```diff
@@ -8,24 +8,44 @@
...
+			endif()
+			separate_arguments(CUDA_TARGET_COMPUTE_CAPABILITY_ UNIX_COMMAND "${CUDA_TARGET_COMPUTE_CAPABILITY_}")
+			list(GET CUDA_TARGET_COMPUTE_CAPABILITY_ 1 FORMATTED_COMPUTE_CAPABILITY)
+			list(GET CUDA_TARGET_COMPUTE_CAPABILITY_ 0 CUDA_TARGET_COMPUTE_CAPABILITY_)
+		endif()
 	else()
-		set(CUDA_TARGET_COMPUTE_CAPABILITY_ $ENV{CUDA_SM})
-	endif()
+		if("$ENV{CUDA_SM}" STREQUAL "")
+			set(ENV{CUDA_INCLUDE_DIRS} "${CUDA_INCLUDE_DIRS}")
```
Comment (eyalroz) [inline cmake/Modules/HandleCUDAComputeCapability.cmake]: Again, there are internal CMake variables for that.

File: examples/by_runtime_api_module/device_management.cpp [hunk 5]
```diff
@@ -62,7 +62,7 @@ int main(int argc, char **argv)
 	// Attributes and properties
 	// ---------------------------
 
-	auto max_registers_per_block = device.get_attribute(cudaDevAttrMaxRegistersPerBlock);
+	auto max_registers_per_block = device.get_attribute(static_cast<cuda::device::attribute_t>(cudaDevAttrMaxRegistersPerBlock));
```
Comment (eyalroz) [inline examples/by_runtime_api_module/device_management.cpp]: Ah, it looks like the return type of get_attribute needs to change, there shouldn't be a cast here.

File: examples/check_cuda.cu [hunk 6]
```diff
@@ -0,0 +1,26 @@
+#include <stdio.h>
```
Comment (eyalroz) [inline examples/check_cuda.cu]: Again, move this to a separate PR.

File: examples/modified_cuda_samples/vectorAdd/vectorAdd.cu [hunk 7]
```diff
@@ -59,7 +59,7 @@ int main(void)
 
 	cuda::launch(
 		vectorAdd,
-		{ blocksPerGrid, threadsPerBlock },
+		{ cuda::grid_block_dimensions_t(blocksPerGrid), cuda::grid_block_dimensions_t(threadsPerBlock) },
```
Comment (eyalroz) [inline examples/modified_cuda_samples/vectorAdd/vectorAdd.cu]: 1. Why did you need this?
2. The two parameters have different types.

File: src/cuda/api/device.hpp [hunk 8]
```diff
@@ -440,9 +440,9 @@ class device_t {
 	 */
 	device::pci_location_t pci_id() const
 	{
-		auto pci_domain_id = get_attribute(cudaDevAttrPciDomainId);
-		auto pci_bus_id = get_attribute(cudaDevAttrPciBusId);
-		auto pci_device_id = get_attribute(cudaDevAttrPciDeviceId);
+		auto pci_domain_id = get_attribute(static_cast<attribute_t>(cudaDevAttrPciDomainId));
```
Comment (eyalroz) [inline src/cuda/api/device.hpp]: Again, need to look at the `get_attribute()` function itself rather than cast its results all the time.

File: examples/by_runtime_api_module/device_management.cpp (hunk 5 above)
Comment (jvschroeder) [reply examples/by_runtime_api_module/device_management.cpp]: I suspect as much, but since I currently don't have the opportunity to test this on a non-Windows machine I don't know if changing the return type would be ok.

File: examples/modified_cuda_samples/vectorAdd/vectorAdd.cu (hunk 7 above)
Comment (jvschroeder) [reply examples/modified_cuda_samples/vectorAdd/vectorAdd.cu]: If I don't have this then I get the following errors

> ./cuda-api-wrappers/examples/modified_cuda_samples/vectorAdd/vectorAdd.cu(60): error C23
> 97: conversion from 'int' to 'cuda::grid_dimensions_t' requires a narrowing conversion [...]
> ./cuda-api-wrappers/examples/modified_cuda_samples/vectorAdd/vectorAdd.cu(60): error C23
> 97: conversion from 'int' to 'cuda::grid_block_dimensions_t' requires a narrowing conversion [...]

File: cmake/Modules/HandleCUDAComputeCapability.cmake (hunk 4 above)
Comment (jvschroeder) [reply cmake/Modules/HandleCUDAComputeCapability.cmake]: set(ENV stuff is from the current (unix) version of the module

File: cmake/Modules/HandleCUDAComputeCapability.cmake (hunk 3 above)
Comment (jvschroeder) [reply cmake/Modules/HandleCUDAComputeCapability.cmake]: regarding 3. - I do not see a reason why we couldn't use the CMake version on Unix as well. - I simply didn't delete it and made the Windows stuff conditional since I can't test if it has exactly the same behaviour.

File: CMakeLists.txt (hunk 1 above)
Comment (jvschroeder) [reply CMakeLists.txt]: Ok - I'll make the change conditional then.

**Comment by jvschroeder** [comment]
First of all sorry for creating a pull-request on the wrong branch. I changed the target branch to development.
I am not sure what will happen to your comments when I split this PR into different branches, so I'll comment on your questions / requests before I make the split you requested.

**Comment by eyalroz** [comment]
Also, @jvschroeder : When you notice a problem, I'd like to ask you to first create an issue on the main repo's issues page. The reason is that sometimes (like in the case of some of the casts) the appropriate fix is more general than the work-around you might write. Also, some other followers might have what to say as well (which is less likely for PRs).
